[feat] Automatically create a Spack environment when not set

giordano opened the pull request:
Fix #2028.

giordano commented on reframe/core/buildsystems.py:
```diff
@@ -826,7 +826,8 @@ class Spack(BuildSystem):
     #: :attr:`~reframe.core.pipeline.RegressionTest.sourcesdir`.
     #:
-    #: This field is required.
+    #: If this field is `None`, the default, the environment name will
+    #: be automatically set to `rfm_spack_env`.
```
I left the reference to `None` because it was already there, but I'm not sure that's accurate:
https://github.com/eth-cscs/reframe/blob/24321bb7e5a9f7a7e4e4486882d326c6809cad0f/reframe/core/buildsystems.py#L833
https://github.com/eth-cscs/reframe/blob/24321bb7e5a9f7a7e4e4486882d326c6809cad0f/reframe/core/buildsystems.py#L866

vkarak replied:
Actually, this can be stale. I would change it to `None`. @rsarm does it ring a bell to you?

giordano replied:
This was changed in #2095

giordano commented on reframe/core/buildsystems.py:
```diff
@@ -826,7 +826,8 @@ class Spack(BuildSystem):
     #: ReFrame looks for environments in the test's
     #: :attr:`~reframe.core.pipeline.RegressionTest.sourcesdir`.
```
Also, I don't think this is true, as far as I can tell the environment is looked in the stage directory

vkarak replied:
The contents of the `sourcesdir` are copied over to the stage directory before entering the compile stage. The environment is a test "resource", so it has to reside in the `sourcesdir` if given.

giordano replied:
```console
% spack env create -d tutorials/basics/stream/spack-env
==> Updating view at /home/mose/repo/reframe/tutorials/basics/stream/spack-env/.spack-env/view
==> Created environment in /home/mose/repo/reframe/tutorials/basics/stream/spack-env
==> You can activate this environment with:
==>   spack env activate /home/mose/repo/reframe/tutorials/basics/stream/spack-env
% ./bin/reframe -c tutorials/basics/stream/stream1.py -r --performance-report
[ReFrame Setup]
  version:           3.7.0-dev.2+1bc108c6
  command:           './bin/reframe -c tutorials/basics/stream/stream1.py -r --performance-report'
  launched by:       mose@dream
  working directory: '/home/mose/repo/reframe'
  settings file:     '/home/mose/repo/reframe/tutorials/config/mysettings.py'
  check search path: '/home/mose/repo/reframe/tutorials/basics/stream/stream1.py'
  stage directory:   '/home/mose/repo/reframe/stage'
  output directory:  '/home/mose/repo/reframe/output'

[==========] Running 1 check(s)
[==========] Started on Thu Jun 24 23:28:26 2021 

[----------] started processing Test (Test)
[ RUN      ] Test on catalina:default using gnu
[     FAIL ] (1/1) Test on catalina:default using gnu [compile: 0.784s run: n/a total: 0.793s]
==> test failed during 'compile_wait': test staged in '/home/mose/repo/reframe/stage/catalina/default/gnu/Test'
[----------] finished processing Test (Test)

[----------] waiting for spawned checks to finish
[----------] all spawned checks have finished

[  FAILED  ] Ran 1/1 test case(s) from 1 check(s) (1 failure(s), 0 skipped)
[==========] Finished on Thu Jun 24 23:28:27 2021 

==============================================================================
SUMMARY OF FAILURES
------------------------------------------------------------------------------
FAILURE INFO for Test 
  * Test Description: Test
  * System partition: catalina:default
  * Environment: gnu
  * Stage directory: /home/mose/repo/reframe/stage/catalina/default/gnu/Test
  * Node list: 
  * Job type: local (id=None)
  * Dependencies (conceptual): []
  * Dependencies (actual): []
  * Maintainers: []
  * Failing phase: compile_wait
  * Rerun with '-n Test -p gnu --system catalina:default -r'
  * Reason: build error: stdout: 'rfm_Test_build.out', stderr: 'rfm_Test_build.err'
--- rfm_Test_build.err (first 10 lines) ---
==> Error: No such environment: 'spack-env'
==> Error: `spack add` requires an environment
  activate an environment first:
      spack env activate ENV
  or use:
      spack -e ENV add ...
--- rfm_Test_build.err --- 
------------------------------------------------------------------------------

Log file(s) saved in: '/tmp/rfm-rg3uuekm.log'
```
Either I'm doing something wrong (totally possible but documentation would help :slightly_smiling_face:) or that doesn't work

giordano replied:
Oooh, sorry, now I see: I need to set `sourcesdir` in the test.  Well, I think this is one more reason for this pull request :slightly_smiling_face:

vkarak commented on reframe/core/buildsystems.py:
```diff
@@ -886,8 +885,12 @@ def emit_build_commands(self, environ):
+        cmds = ['. $SPACK_ROOT/share/spack/setup-env.sh']
+        if self.environment is _Undefined:
+            self.environment = 'rfm_spack_env'
+            cmds.append(f'spack env create -d {self.environment}')
```
Leave a blank line after this one.

vkarak replied:
Sorry I meant before this one... I can fix that, no worries.

vkarak reviewed (commented):
I tried this PR with the following test and it works:

```python
import reframe as rfm
import reframe.utility.sanity as sn


@rfm.simple_test
class reframe_spack_env_check(rfm.RegressionTest):
    valid_systems = ['*']
    valid_prog_environs = ['*']
    build_system = 'Spack'
    executable = 'bzip2'
    executable_opts = ['--help']

    @run_before('sanity')
    def set_sanity(self):
        self.sanity_patterns = sn.assert_found(r'Version 1.0.7', self.stderr)

    @run_before('compile')
    def setup_build_system(self):
        self.build_system.specs = ['bzip2@1.0.7']
```

However, when I run the test and then do `spack find bzip2`, I see 1.0.7 installed in my system. Is this expected? I'm not that familiar with Spack, so I might miss something. @victorusu can you also have a look?

vkarak reviewed (changes requested):
So the problem with the automatically generated environment is that it doesn't set correctly the installation paths. Ideally, we want this to be installed in the stage directory of the test and not "pollute" the user's installation, unless explicitly required.

giordano commented:
How is this different from an environment manually created from the user?  I guess I'm missing something, but then we're back to the fact documentation isn't helpful at the moment and I'm trying to reverse engineer what I'm supposed to do :slightly_smiling_face:

victorusu commented:
@giordano, I think the issue is that we haven't documented enough the current goals of this feature.

AFAIK, what we wanted in the beginning is to test Spack packages and have them installed in test's `stagedir`. Then, we would see how it works in the wild and evolve it to achieve other things.
Therefore, we left the environments out-of-the-box to have this discussion later and told people to enable the environment before starting reframe. In this case, the user can set the environment as they want and still leverage the feature from reframe.

One of the reasons for this decision was that we did not know, in principle, which options of the `Spack.yaml` file we would need to add to the build system. 
So we were waiting for two things:
1. To have feedback from he Community
2. As you know Spack evolves very fast and so we would like to wait a bit until environments also become more and more stable in Spack.

So, we postponed environment's inclusion/discussion.

Now, as you have pointed out, the time to have this discussion has come!😄 

@vkarak and @giordano, what do you think about adding a build system option to set the `install_dir`?
In this case, it would create on the stage dir a config file from Spack that would set the `install_dir` and we could make it point to the `stagedir`, by default?

giordano commented:
To be clear, what you're suggesting is to automatically run something like
```
spack config add "config:install_tree:${install_dir}"
```
after activating the environment, where `${install_dir}` is the chosen directory, something like `${stagedir}/prefix` by default?  That'd be fine with me as long as there is a reasonable default which doesn't require the user to do anything extra: as a user and someone who will need to tell other people how to use this, I wouldn't be particularly happy about _having_ to manually create the environment with the right configuration, maybe outside of reframe.

I'd love to have something as simple as this
```python
import reframe as rfm
import reframe.utility.sanity as sn


@rfm.simple_test
class Test(rfm.RegressionTest):
    valid_systems = ['*']
    valid_prog_environs = ['*']
    build_system = 'Spack'
    executable = 'myprogram'

    @run_before('compile')
    def set_compiler_flags(self):
        self.build_system.specs = ['program@1.0']

    @run_before('sanity')
    def set_sanity_patterns(self):
        self.sanity_patterns = sn.assert_found(r'SUCCESS', self.stdout)
```
to work without anything extra needed.  As far as I understand, at the moment this isn't possible, and that's why I'm trying to hack the Spack build system with this pull request :slightly_smiling_face:

victorusu commented:
> To be clear, what you're suggesting is to automatically run something like
> 
> ```
> spack config add "config:install_tree:${install_dir}"
> ```
> 
> after activating the environment, where `${install_dir}` is the chosen directory, something like `${stagedir}/prefix` by default? That'd be fine with me as long as there is a reasonable default which doesn't require the user to do anything extra: as a user and someone who will need to tell other people how to use this, I wouldn't be particularly happy about _having_ to manually create the environment with the right configuration, maybe outside of reframe.
> 
> I'd love to have something as simple as this
> 
> ```python
> import reframe as rfm
> import reframe.utility.sanity as sn
> 
> 
> @rfm.simple_test
> class Test(rfm.RegressionTest):
>     valid_systems = ['*']
>     valid_prog_environs = ['*']
>     build_system = 'Spack'
>     executable = 'myprogram'
> 
>     @run_before('compile')
>     def set_compiler_flags(self):
>         self.build_system.specs = ['program@1.0']
> 
>     @run_before('sanity')
>     def set_sanity_patterns(self):
>         self.sanity_patterns = sn.assert_found(r'SUCCESS', self.stdout)
> ```
> 
> to work without anything extra needed. As far as I understand, at the moment this isn't possible, and that's why I'm trying to hack the Spack build system with this pull request 🙂

@giordano, that's exactly what I am saying. @vkarak what do you think?

vkarak commented:
@giordano Yes, the rationale from @victorusu is correct about the initial purpose of the feature. We have been seeing it also as feature that would help in CI or to do parameter exploration on different builds. We never thought about it seriously as a means of deployment, because there can be too many parameters to consider there. Also I totally agree with what you'd like to see from a user's perspective. My initial objection was that, as a user, I would not like to see my global environment to be implicitly modified after running a reframe test. So by default the `install_dir` of the environment you create should point to the stage dir.

@victorusu For this PR, I would say that emitting the `spack config add ...` should be enough. If users want to customize the environment, they can simply create one by themselver for the moment. As far as I can see if `spack config` is done inside the environment then it's scoped in that. So that should be fine, imho.

giordano commented:
Ok, I added an `install_tree` (used same name as Spack) to the build system.  This can be either a non-empty string or `None`:
* if `environment` and `install_tree` are both `None`, then `install_tree` is automatically set to `install_tree` and this is used to set Spack's install tree
* if `environment` is not `None` but `install_tree` is, nothing happens (I feel like if the user provides their own environment, they're going to configure it themselves)

vkarak commented on reframe/core/buildsystems.py:
```diff
@@ -865,16 +885,19 @@ def __init__(self):
 
         ret = self._env_activate_cmds()
+        if self.install_tree:
+            ret.append('spack config add '
```
Just for symmetry:

```suggestion
            ret.append(f'spack config add '
```

vkarak commented on reframe/core/buildsystems.py:
```diff
@@ -887,8 +910,15 @@ def emit_build_commands(self, environ):
+        if self.environment is _Undefined:
+            self.environment = 'rfm_spack_env'
+            if self.install_tree is None:
+                self.install_tree = 'install_tree'
```
Also I would not set the `install_tree` like that but rather do

```python
install_tree = self.install_tree or 'opt/spack'
```

And then use `install_tree`. The reason for that is that we don't want to change the stage of the object: user had it as `None` and then it magically becomes `opt/spack`.

giordano commented:
Ok, I finally got back to this pull request, rebased on latest `master`, and addressed all comments.  Main relevant change now is that the default value of `environment` and `install_tree` is an empty string (which is a bit like going the opposite direction of #2095) :slightly_smiling_face:

vkarak reviewed (changes requested):
Lgtm. I just have some minor suggestions. I could also have a look at the tutorial to see if it needs to be adjusted.

pep8speaks commented:
Hello @giordano, Thank you for updating!

* In the file [`reframe/core/buildsystems.py`](https://github.com/eth-cscs/reframe/blob/58e5c617b96a8246e8fd8155c6c277168794ce5d/reframe/core/buildsystems.py):

> [Line 807:80](https://github.com/eth-cscs/reframe/blob/58e5c617b96a8246e8fd8155c6c277168794ce5d/reframe/core/buildsystems.py#L807): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 808:80](https://github.com/eth-cscs/reframe/blob/58e5c617b96a8246e8fd8155c6c277168794ce5d/reframe/core/buildsystems.py#L808): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)



Do see the ReFrame [Coding Style Guide](https://github.com/eth-cscs/reframe/wiki/coding-style-guide)

giordano commented:
I mentioned in the tutorial that the environment isn't required and it'll be automatically created by ReFrame if not provided.  I didn't mention `install_tree` there, do you think it's necessary?

vkarak reviewed (approved):
> I mentioned in the tutorial that the environment isn't required and it'll be automatically created by ReFrame if not provided. I didn't mention install_tree there, do you think it's necessary?

No, it's not necessary. It's enough the we mention it in the reference.

Lgtm now! Going in.

vkarak commented:
ok to test

giordano commented:
Thanks!
